### PR TITLE
storage: Support adding vector index in background

### PR DESCRIPTION
**Pull request by Lloyd-Pottiger**
### What problem does this PR solve?

Issue Number: ref #9032

Problem Summary:

### What is changed and how it works?

```commit-message
storage: Support adding vector index in background
```

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [x] Breaking backward compatibility
  - STORAGE_FORMAT_CURRENT  is bumped to `STORAGE_FORMAT_V7`

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: dbms/src/Server/Server.cpp [hunk 1]
```diff
@@ -1325,6 +1326,24 @@ int Server::main(const std::vector<std::string> & /*args*/)
         settings.max_memory_usage_for_all_queries.getActualBytes(server_info.memory_info.capacity),
         settings.bytes_that_rss_larger_than_limit);
 
+    if (global_context->getSharedContextDisagg()->isDisaggregatedComputeMode())
+    {
+        // No need to have local index scheduler.
+    }
+    else if (global_context->getSharedContextDisagg()->isDisaggregatedStorageMode())
+    {
+        global_context->initializeGlobalLocalIndexerScheduler(
```
Comment (CalvinNeo) [inline dbms/src/Server/Server.cpp:1335]: Can you explain this math?

Comment (Lloyd-Pottiger) [reply dbms/src/Server/Server.cpp:1335]: initialize the LocalIndexerScheduler introduce by https://github.com/pingcap/tiflash/pull/9383, all vector index build task will be schedule by this scheduler.

Comment (CalvinNeo) [reply dbms/src/Server/Server.cpp:1335]: I checked that PR, and I can not figure out why it is `* 8 / 10`

Comment (Lloyd-Pottiger) [reply dbms/src/Server/Server.cpp:1338]:
```
// There is no compute task in write node.
// Set the pool size to 80% of logical cores and 60% of memory
// to take full advantage of the resources and avoid blocking other tasks like writes and compactions.
```

Add some comments.

File: dbms/src/Storages/DeltaMerge/DeltaMergeStore_InternalSegment.cpp [hunk 2]
```diff
@@ -389,9 +441,293 @@ SegmentPtr DeltaMergeStore::segmentMerge(
...
+
+    size_t segments_updated_meta = 0;
+    auto dm_context = newDMContext(global_context, global_context.getSettingsRef(), "checkAllSegmentsLocalIndex");
+
+    // 1. Make all segments referencing latest meta version.
+    {
+        Stopwatch watch;
+        std::unique_lock lock(read_write_mutex);
+
+        std::map<PageIdU64, DMFilePtr> latest_dmf_by_id;
```
Comment (CalvinNeo) [inline dbms/src/Storages/DeltaMerge/DeltaMergeStore_InternalSegment.cpp:463]: Where is `latest_dmf_by_id` filled?

File: dbms/src/Storages/DeltaMerge/DeltaMergeStore_InternalSegment.cpp [hunk 3]
```diff
@@ -389,9 +441,293 @@ SegmentPtr DeltaMergeStore::segmentMerge(
...
+        "CheckAllSegmentsLocalIndex - Finish, segments_[updated_meta/missing_index]={}/{}",
+        segments_updated_meta,
+        segments_missing_indexes);
+}
+
+bool DeltaMergeStore::segmentEnsureStableIndexAsync(const SegmentPtr & segment)
+{
+    RUNTIME_CHECK(segment != nullptr);
+
+    // TODO(local index): There could be some indexes are built while some indexes is not yet built after DDL
```
Comment (CalvinNeo) [inline dbms/src/Storages/DeltaMerge/DeltaMergeStore_InternalSegment.cpp:526]: There could be some indexes are **being** built?

Comment (Lloyd-Pottiger) [reply dbms/src/Storages/DeltaMerge/DeltaMergeStore_InternalSegment.cpp:526]: Just ignore it, will remove it in a later PR.

File: dbms/src/Storages/DeltaMerge/DeltaMergeStore_InternalSegment.cpp [hunk 4]
```diff
@@ -389,9 +441,293 @@ SegmentPtr DeltaMergeStore::segmentMerge(
...
+    // So after TiFlash restarts, we will update meta versions to latest versions again.
+    {
+        // We must acquire a single lock when updating multiple segments.
+        // Otherwise we may miss new segments.
+        std::unique_lock lock(read_write_mutex);
+        auto segment_ids = dmfile_id_to_segment_ids.get(dm_file->fileId());
+        for (const auto & seg_id : segment_ids)
+        {
+            auto segment = id_to_segment[seg_id];
+            auto new_segment = segmentUpdateMeta(lock, dm_context, segment, new_dmfiles);
```
Comment (CalvinNeo) [inline dbms/src/Storages/DeltaMerge/DeltaMergeStore_InternalSegment.cpp:724]: In segmentUpdateMeta, we will acquire a lock for `using Lock = std::unique_lock<std::recursive_mutex>;` of a DeltaValueSpace for a Segment, while we are holding also a unique_lock of DM.
I wonder if the lock is too grain? AFAIK, the segment lock is only used in `replaceStableMetaVersion`. So I wonder if we could do `replaceStableMetaVersion` without DM lock?

Comment (Lloyd-Pottiger) [reply dbms/src/Storages/DeltaMerge/DeltaMergeStore_InternalSegment.cpp:724]: Accquiring DM lock to make sure segment will not be abandoned by other thread.

File: dbms/src/Storages/FormatVersion.h [hunk 5]
```diff
@@ -181,7 +181,7 @@ inline static const StorageFormatVersion STORAGE_FORMAT_V102 = StorageFormatVers
     .identifier = 102,
 };
 
-inline StorageFormatVersion STORAGE_FORMAT_CURRENT = STORAGE_FORMAT_V5;
+inline StorageFormatVersion STORAGE_FORMAT_CURRENT = STORAGE_FORMAT_V6;
```
Comment (JaySon-Huang) [inline dbms/src/Storages/FormatVersion.h:184]:
```suggestion
inline StorageFormatVersion STORAGE_FORMAT_CURRENT = STORAGE_FORMAT_V7;
```

File: dbms/src/Server/Server.cpp [hunk 6]
```diff
@@ -987,12 +987,13 @@ int Server::main(const std::vector<std::string> & /*args*/)
 
     if (storage_config.format_version != 0)
     {
-        if (storage_config.s3_config.isS3Enabled() && storage_config.format_version != STORAGE_FORMAT_V100.identifier)
+        if (storage_config.s3_config.isS3Enabled() && storage_config.format_version != STORAGE_FORMAT_V100.identifier
+            && storage_config.format_version != STORAGE_FORMAT_V101.identifier)
         {
-            LOG_WARNING(log, "'storage.format_version' must be set to 100 when S3 is enabled!");
+            LOG_WARNING(log, "'storage.format_version' must be set to 100 or 101 when S3 is enabled!");
```
Comment (JaySon-Huang) [inline dbms/src/Server/Server.cpp:993]:
```suggestion
            LOG_WARNING(log, "'storage.format_version' must be set to 100/101/102 when S3 is enabled!");
```

File: dbms/src/Server/Server.cpp [hunk 7]
```diff
@@ -987,12 +987,13 @@ int Server::main(const std::vector<std::string> & /*args*/)
 
     if (storage_config.format_version != 0)
     {
-        if (storage_config.s3_config.isS3Enabled() && storage_config.format_version != STORAGE_FORMAT_V100.identifier)
+        if (storage_config.s3_config.isS3Enabled() && storage_config.format_version != STORAGE_FORMAT_V100.identifier
+            && storage_config.format_version != STORAGE_FORMAT_V101.identifier)
```
Comment (JaySon-Huang) [inline dbms/src/Server/Server.cpp:991]:
```suggestion
        if (storage_config.s3_config.isS3Enabled() && storage_config.format_version != STORAGE_FORMAT_V100.identifier
            && storage_config.format_version != STORAGE_FORMAT_V101.identifier
            && storage_config.format_version != STORAGE_FORMAT_V102.identifier)
```

File: dbms/src/Server/Server.cpp [hunk 8]
```diff
@@ -987,12 +987,13 @@ int Server::main(const std::vector<std::string> & /*args*/)
...
         {
-            LOG_WARNING(log, "'storage.format_version' must be set to 100 when S3 is enabled!");
+            LOG_WARNING(log, "'storage.format_version' must be set to 100 or 101 when S3 is enabled!");
             throw Exception(
                 ErrorCodes::INVALID_CONFIG_PARAMETER,
-                "'storage.format_version' must be set to 100 when S3 is enabled!");
+                "'storage.format_version' must be set to 100 or 101 when S3 is enabled!");
```
Comment (JaySon-Huang) [inline dbms/src/Server/Server.cpp:996]:
```suggestion
                "'storage.format_version' must be set to 100/101/102 when S3 is enabled!");
```

File: dbms/src/Server/Server.cpp [hunk 9]
```diff
@@ -1003,8 +1004,8 @@ int Server::main(const std::vector<std::string> & /*args*/)
         {
             // If the user does not explicitly set format_version in the config file but
             // enables S3, then we set up a proper format version to support S3.
-            setStorageFormat(STORAGE_FORMAT_V100.identifier);
-            LOG_INFO(log, "Using format_version={} (infer by S3 is enabled).", STORAGE_FORMAT_V100.identifier);
+            setStorageFormat(STORAGE_FORMAT_V101.identifier);
+            LOG_INFO(log, "Using format_version={} (infer by S3 is enabled).", STORAGE_FORMAT_V101.identifier);
```
Comment (JaySon-Huang) [inline dbms/src/Server/Server.cpp:1008]:
```suggestion
            setStorageFormat(STORAGE_FORMAT_V102.identifier);
            LOG_INFO(log, "Using format_version={} (infer by S3 is enabled).", STORAGE_FORMAT_V102.identifier);
```

**Review by JaySon-Huang — COMMENTED** [review]
Directly bump the version for supporting building index on both DeltaVS and StableVS

File: dbms/src/Server/Server.cpp [hunk 10]
```diff
@@ -1325,6 +1326,27 @@ int Server::main(const std::vector<std::string> & /*args*/)
...
+    {
+        // No need to have local index scheduler.
+    }
+    else if (global_context->getSharedContextDisagg()->isDisaggregatedStorageMode())
+    {
+        // There is no compute task in write node.
+        // Set the pool size to 80% of logical cores and 60% of memory
+        // to take full advantage of the resources and avoid blocking other tasks like writes and compactions.
+        global_context->initializeGlobalLocalIndexerScheduler(
+            server_info.cpu_info.logical_cores * 8 / 10,
```
Comment (JaySon-Huang) [inline dbms/src/Server/Server.cpp:1339]: Add a protect `std::max(1, ...)` for the pool_size?

File: dbms/src/Server/Server.cpp [hunk 11]
```diff
@@ -1325,6 +1326,27 @@ int Server::main(const std::vector<std::string> & /*args*/)
...
+            server_info.memory_info.capacity * 6 / 10);
+    }
+    else
+    {
+        // There could be compute tasks, reserve more memory for computes.
+        global_context->initializeGlobalLocalIndexerScheduler(
+            server_info.cpu_info.logical_cores * 4 / 10,
```
Comment (JaySon-Huang) [inline dbms/src/Server/Server.cpp:1346]: ditto

File: dbms/src/Storages/DeltaMerge/File/DMFileV3IncrementWriter.cpp [hunk 12]
```diff
@@ -148,15 +146,6 @@ void DMFileV3IncrementWriter::writeAndIncludeMetaFile()
...
     auto meta_file_path_for_write = meta_file_path + ".tmp";
 
-    // Just a protection. We don't allow overwriting meta file.
-    {
-        auto existing_file = Poco::File(meta_file_path);
-        RUNTIME_CHECK_MSG( //
-            !existing_file.exists(),
-            "Meta file already exists, file={}",
-            meta_file_path);
-    }
```
Comment (JaySon-Huang) [inline dbms/src/Storages/DeltaMerge/File/DMFileV3IncrementWriter.cpp:158]: why these lines are removed?

**Review by breezewish — APPROVED** [review]
The change looks minimal compared as CSE version.

File: dbms/src/Server/Server.cpp (hunk 10 above)
Comment (Lloyd-Pottiger) [reply dbms/src/Server/Server.cpp]: done

File: dbms/src/Storages/DeltaMerge/DeltaMergeStore_InternalSegment.cpp [hunk 13]
```diff
@@ -389,9 +441,293 @@ SegmentPtr DeltaMergeStore::segmentMerge(
...
+                    // Note: pageId could be different. It is fine, replaceStableMetaVersion will fix it.
+                    auto update_result = segmentUpdateMeta(lock, *dm_context, segment, {latest_dmf});
+                    RUNTIME_CHECK(update_result != nullptr, segment->simpleInfo());
+                    ++segments_updated_meta;
+                }
+            }
+        }
+        LOG_INFO(
+            log,
+            "CheckAllSegmentsLocalIndex - Finish, updated_meta={}, elapsed={}",
```
Comment (JaySon-Huang) [inline dbms/src/Storages/DeltaMerge/DeltaMergeStore_InternalSegment.cpp:492]:
```suggestion
            "CheckAllSegmentsLocalIndex - Finish, updated_meta={}, elapsed={:.3f}s",
```

File: dbms/src/Storages/DeltaMerge/File/DMFileV3IncrementWriter.cpp (hunk 12 above)
Comment (Lloyd-Pottiger) [reply dbms/src/Storages/DeltaMerge/File/DMFileV3IncrementWriter.cpp:158]: The meta file may be broken.

File: dbms/src/Storages/DeltaMerge/Segment.cpp [hunk 14]
```diff
@@ -1470,8 +1465,11 @@ SegmentPtr Segment::replaceStableMetaVersion(
 
     wbs.writeAll();
 
-    LOG_DEBUG(log, "ReplaceStableMetaVersion - Finish, new_stable_files={}", new_stable_files_str());
-
+    LOG_DEBUG(
```
[the remark this answers is not in the document]
Comment (Lloyd-Pottiger) [reply dbms/src/Storages/DeltaMerge/Segment.cpp:1468]: We will print info log in `DeltaMergeStore::segmentUpdateMeta`

**Review by JaySon-Huang — APPROVED** [review]
LGTM

**Pull request merged**